pkg/store/proxy_heap.go

```diff
@@ -0,0 +1,261 @@
+// Copyright (c) The Thanos Authors.
+// Licensed under the Apache License 2.0.
+
+package store
+
+import (
+	"container/heap"
+	"sort"
+
+	"github.com/prometheus/prometheus/model/labels"
+	"github.com/prometheus/prometheus/storage"
+	"github.com/thanos-io/thanos/pkg/store/labelpb"
+	"github.com/thanos-io/thanos/pkg/store/storepb"
+)
+
+// dedupResponseHeap is a wrapper around ProxyResponseHeap
+// that deduplicates identical chunks identified by the same labelset.
+// It uses a hashing function to do that.
+type dedupResponseHeap struct {
+	h *ProxyResponseHeap
+
+	responses []*storepb.SeriesResponse
+
+	previousResponse *storepb.SeriesResponse
+	previousNext     bool
+}
+
+func NewDedupResponseHeap(h *ProxyResponseHeap) *dedupResponseHeap {
+	return &dedupResponseHeap{
+		h:            h,
+		previousNext: h.Next(),
+	}
+}
+
+func (d *dedupResponseHeap) At() *storepb.SeriesResponse {
+	defer func() {
+		d.responses = d.responses[:0]
+	}()
+	if len(d.responses) == 0 {
+		return nil
+	} else if len(d.responses) == 1 {
+		return d.responses[0]
+	}
+
+	chunkDedupMap := map[string]*storepb.AggrChunk{}
+
+	for _, resp := range d.responses {
+		for _, chk := range resp.GetSeries().Chunks {
+			h := chk.Hash()
+
+			if _, ok := chunkDedupMap[h]; !ok {
```

Comment:
Curious why do we need to do deduplication here? Will exactly the same chunks be returned from the target stores in a single `Series` grpc call?

Reply:
To avoid duplicating work when multiple StoreAPIs return the same chunks. This can happen due to misconfiguration. For example, multiple Thanos Store instances connected that use the same bucket.